mqtt/mqtt-broker/src/sys_topic.rs

```diff
@@ -0,0 +1,399 @@
+
+        // One session
+        let sessions: HashMap<ClientId, Session> = (1..2)
+            .map(|i| {
+                let id = format!("Session {}", i);
+                let session =
+                    make_session(&id, (1..i).map(|j| format!("Subscription {}", j)), true);
+
+                (id.into(), session)
+            })
+            .collect();
+        let one_connection = StateChange::new_connection(&sessions);
+        if let StateChange::Connections(one_connection) = &one_connection {
+            let expected: ClientId = "Session 1".into();
+            assert_eq!(&vec![&expected], one_connection);
+        }
+
+        let message: proto::Publication = one_connection.try_into().unwrap();
+        matches_connection_publication(message, &["Session 1"]);
+
+        // Multiple sessions
+        let sessions: HashMap<ClientId, Session> = (1..4)
+            .map(|i| {
+                let id = format!("Session {}", i);
+                let session =
+                    make_session(&id, (1..i).map(|j| format!("Subscription {}", j)), true);
+
+                (id.into(), session)
+            })
+            .collect();
+        let many_connections = StateChange::new_connection(&sessions);
+        if let StateChange::Connections(many_connections) = &many_connections {
+            let mut actual: Vec<&str> = many_connections.iter().map(|id| id.as_str()).collect();
+            actual.sort();
+            assert_eq!(vec!["Session 1", "Session 2", "Session 3"], actual);
+        }
+
+        let message: proto::Publication = many_connections.try_into().unwrap();
+        matches_connection_publication(message, &["Session 1", "Session 2", "Session 3"]);
+
+        // Offline sessions
+        let sessions: HashMap<ClientId, Session> = (1..7)
+            .map(|i| {
+                let id = format!("Session {}", i);
+                let session = make_session(
+                    &id,
+                    (1..i).map(|j| format!("Subscription {}", j)),
+                    i % 2 == 0, // even sessions are online, odd offline
+                );
+
+                (id.into(), session)
+            })
+            .collect();
+        let many_connections = StateChange::new_connection(&sessions);
+        if let StateChange::Connections(many_connections) = &many_connections {
+            let mut actual: Vec<&str> = many_connections.iter().map(|id| id.as_str()).collect();
+            actual.sort();
+            assert_eq!(vec!["Session 2", "Session 4", "Session 6"], actual);
+        }
+
+        let message: proto::Publication = many_connections.try_into().unwrap();
+        matches_connection_publication(message, &["Session 2", "Session 4", "Session 6"]);
+    }
+
+    #[test]
+    fn test_sessions() {
+        // No sessions
+        let sessions: HashMap<ClientId, Session> = HashMap::new();
+        let no_sessions = StateChange::new_session(&sessions);
+        if let StateChange::Connections(no_sessions) = &no_sessions {
```

Comment:
Here and further in this test it should be `StateChange::Sessions(no_sessions)`. I would recommend try to avoid any `if` in the tests (otherwise who will test the correctness of the test? :) ). Instead, use assert_matches! if possible.

Reply:
that works for this test, but for similar tests further down I need to sort the returned vec, which a pattern match cannot do (afaik)

Reply:
Sure, `assert_matches!` might not work everywhere, use `assert_eq!` then. And don't use `if`.
This line should look like below. Please note that `StateChange::new_session()` returns `StateChange::Sessions`, and not `StateChange::Connections`.
```
        assert_eq!(
            no_sessions,
            StateChange::Sessions(Vec::<&ClientId>::new())
        );
```